### PR TITLE
Hw 04

### PR DESCRIPTION
**Pull request by AliokY**

### CODE REVIEW

File: HW_3/HW03.Alphabet/HW03.Alphabet/bin/Debug/net5.0/HW03.Alphabet.deps.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    }
+  },
+  "libraries": {
+    "HW03.Alphabet/1.0.0": {
+      "type": "project",
+      "serviceable": false,
+      "sha512": ""
+    }
+  }
+}
```
Comment (VadzimPapko) [inline HW_3/HW03.Alphabet/HW03.Alphabet/bin/Debug/net5.0/HW03.Alphabet.deps.json:23]: many files - should be in .gitignore
image

File: HW_4/HW04/HW04.Task1/Program.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            Console.WriteLine("Output inputArray");
+            PrintArray(inputArray);
+
+            Console.WriteLine("Output resultArray");
+            PrintArray(resultArray);
+
+            Console.ReadKey();
+        }
+    }
+}
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task1/Program.cs:63]: correct

File: HW_4/HW04/HW04.Task3/Program.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            sw.Start();
+            Array.Reverse(randomArray);
+            sw.Stop();
+            Console.Write("Runtime of Array.Reverse method: ");
+            Console.WriteLine(sw.Elapsed.TotalMilliseconds + " ms");
+
+            Console.ReadKey();
+        }
+    }
+}
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task3/Program.cs:48]: well done

File: HW_4/HW04/HW04.Task4/Program.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                linesVerse[i] = linesVerse[i].Replace("О", "А");
+            }
+            foreach (string line in linesVerse)
+            {
+                Console.WriteLine(line);
+            }
+            Console.ReadKey();
+        }
+    }
+}
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task4/Program.cs:27]: correct

File: HW_4/HW04/HW04.Task5/Program.cs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                if (strArray[i].Length < shortestWord.Length)
+                {
+                    shortestWord = strArray[i];
+                }
+            }
+            int indexLong = Array.IndexOf(strArray, longestWord);
+            int indexShort = Array.IndexOf(strArray, shortestWord);
+
+            strArray[indexLong] = shortestWord;
+            strArray[indexShort] = longestWord;
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task5/Program.cs:76]: correct

File: HW_4/HW04/HW04.Task5/Program.cs [hunk 6]
```diff
@@ -0,0 +1,160 @@
+﻿using System;
+using System.Text.RegularExpressions;
+
+namespace HW04.Task5
+{
+    class Program
+    {
+        private static string DeleteLongestWord(string str)
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task5/Program.cs:8]: possible solution

File: HW_4/HW04/HW04.Task5/Program.cs [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        private static int CountAllSymbols(string str)
+        {
+            str = str.Replace(" ", "");
+
+            int symbolQuantity = 0;
+
+            for (int i = 0; i < str.Length; i++)
+            {
+                // checking for Russian letters and all punctuation marks
+                if (!Regex.IsMatch(str[i].ToString(), @"^[а-яёА-Я]+\p{P}+$"))
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task5/Program.cs:92]: Regex - good one.
also possible to use Char.IsLetter(), Char.IsDigit(), etc.

File: HW_4/HW04/HW04.Task5/Program.cs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+                {
+                    symbolQuantity += 1;
+                }
+            }
+
+            return symbolQuantity;
+        }
+
+        private static int FindLongest(string[] strArray)
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task5/Program.cs:101]: good one

File: HW_4/HW04/HW04.Task5/Program.cs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+            foreach (var item in newArray)
+            {
+                Console.WriteLine(item);
+            }
+
+            Console.ReadKey();
+        }
+    }
+}
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task5/Program.cs:160]: better -put all methods under Main, 

for instance, in the following order

Main
SortString
SelectionSort
FindLongest

File: HW_4/HW04/HW04.Task5/Program.cs [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+            return symbolQuantity;
+        }
+
+        private static int FindLongest(string[] strArray)
+        {
+            string longest = strArray[0];
+            int longestIndex = 0;
+            for (int i = 1; i < strArray.Length; i++)
+            {
+                if (strArray[i].Length > longest.Length)
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task5/Program.cs:107]: Null Reference Exception possible

File: HW_4/HW04/HW04.Task5/Program.cs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        {
+            string[] newArray = strArray;
+
+            string[] newArr = new string[] { };
+            for (int i = 0; i < strArray.Length; i++)
+            {
+                int longestIndex = FindLongest(strArray);
+                newArray[i] = strArray[longestIndex];
+
+                Array.Clear(strArray, longestIndex, 1);
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task5/Program.cs:126]: you need to make sure you cover Null Reference Exception

File: HW_4/HW04/HW04.Task5/Program.cs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+                    longest = strArray[i];
+                    longestIndex = i;
+                }
+            }
+            return longestIndex;
+        }
+
+        private static string[] SelectionSort(string[] strArray)
+        {
+            string[] newArray = strArray;
```
Comment (VadzimPapko) [inline HW_4/HW04/HW04.Task5/Program.cs:118]: now you have two variables pointed to one Array object in HEAP memory - b/c the Array - reference type in C#
so, that means when you change newArray - srtArray also changed and visa versa.

File: HW_4/HW04/Task2/Program.cs [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+            }
+            inputArray[index] = item;
+
+            Console.WriteLine("Output altered array.");
+            PrintArray(inputArray);
+
+            Console.ReadKey();
+        }
+    }
+}
```
Comment (VadzimPapko) [inline HW_4/HW04/Task2/Program.cs:84]: correct

**Review by VadzimPapko — COMMENTED** [review]
Reviewed. Your score is 8

File: HW_4/HW04/HW04.Task5/Program.cs (hunk 11 above)
Comment (AliokY) [reply HW_4/HW04/HW04.Task5/Program.cs:126]: please, could you explain in a little more detail how  I can fix it?

**Pull request merged**